### DISCUSSION TITLE
How to get detailed logs and transaction information for a running Anvil chain?

### DISCUSSION THREAD

**Discussion by wesfloyd** [foundry-rs/foundry#7117] [Help]
When running a local anvil chain I'm able to see basic logs via stdout including block number, block hash, etc. Is there any best practice for me to see more details about the transaction when running anvil locally? Eg
- Is there a local file that I can tail for detailed system logs?
- Are people running a local version of Etherscan to inspect function calls, event emit logs, otherwise?

![image](https://github.com/foundry-rs/foundry/assets/260568/7e580ba4-52dc-4791-b27f-0d0bfac816a7)

I had trouble finding reference to this in their docs, I only found reference to functions like eth_getLogs in their [RPC methods](https://book.getfoundry.sh/reference/anvil/#supported-rpc-methods)

Thank you!

**Comment by ultravigilante** [comment]
This cURL request to the local RPC endpoint might be what you're looking for:

```bash
curl http://localhost:8545 -s -X POST -H "Content-Type: application/json" -d '
  {"jsonrpc":"2.0","id":1, "method":"debug_traceTransaction", "params": ["0x1e3546b098620f7979555c4a21ad4a8ee5d4a6a2a36518177abba2a8a34a8f07"]}
'
```

Replacing `0x1e3546b098620f7979555c4a21ad4a8ee5d4a6a2a36518177abba2a8a34a8f07` with your transaction hash.

Anvil can be started with or without the `--tracing` flag, here's the output of the request for both:

#### With `--tracing`
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "failed": true,
        "gas": 61812,
        "returnValue": "08c379a000000000000[...snip...]000000000000",
        "structLogs": [
            {
                "pc": 0,
                "op": "PUSH1",
                "gas": 277328,
                "gasCost": 3,
                "depth": 1,
                "stack": []
            },
            {
                "pc": 2,
                "op": "PUSH1",
                "gas": 277325,
                "gasCost": 3,
                "depth": 1,
                "stack": [
                    "0x80"
                ]
            },
            {
                "pc": 4,
                "op": "MSTORE",
                "gas": 277322,
                "gasCost": 12,
                "depth": 1,
                "stack": [
                    "0x80",
                    "0x40"
                ]
            },
...
```

The last opcode will be a `REVERT` in the case of errors and it includes the stack values.

#### Without `--tracing`

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "failed": true,
        "gas": 61812,
        "returnValue": "08c379a00000000000[...snip...]00000",
        "structLogs": []
    }
}
```

There's also a `trace_transaction` from the list of RPC calls in Anvil:

```bash
curl http://localhost:8545 -s -X POST -H "Content-Type: application/json" -d '
  {"jsonrpc":"2.0","id":1, "method":"trace_transaction", "params": ["0x1e3546b098620f7979555c4a21ad4a8ee5d4a6a2a36518177abba2a8a34a8f07"]}
'
```

in my case it didn't yield much more information.
